### ISSUE TITLE
Unreasonable type error

### ISSUE THREAD

**Issue by zonyitoo** [carllerche/mio#168]
For the version on crates.io, refer to rust-lang/rust#25443

Check this code snippet

```rust
extern crate mio;

fn main() {
    println!("HELLO?");

    // This should be `mio::tcp::TcpListener`, but rustc insisted that it is `std::net::tcp::TcpListener`
    let listener = mio::tcp::TcpListener::bind("127.0.0.1:1080").unwrap();

    match listener.accept() {
        Ok(None) => {
            println!("WOULDBLOCK");
        },
        Ok(Some(stream)) => {
            println!("Got stream");
        },
        Err(err) => {
            println!("Err {:?}", err);
        }
    }
}
```

In this code snippet should be `mio::tcp::TcpListener`, but rustc insisted that it should be `std::net::tcp::TcpListener`.

Rustc version

```
rustc 1.1.0-nightly (e5394240a 2015-05-14) (built 2015-05-13)
```

Compiler output

```
main.rs:11:12: 11:16 error: mismatched types:
 expected `(std::net::tcp::TcpStream, std::net::addr::SocketAddr)`,
    found `core::option::Option<_>`
(expected tuple,
    found enum `core::option::Option`) [E0308]
main.rs:11         Ok(None) => {
                      ^~~~
main.rs:11:12: 11:16 help: pass `--explain E0308` to see a detailed explanation
main.rs:14:12: 14:24 error: mismatched types:
 expected `(std::net::tcp::TcpStream, std::net::addr::SocketAddr)`,
    found `core::option::Option<_>`
(expected tuple,
    found enum `core::option::Option`) [E0308]
main.rs:14         Ok(Some(stream)) => {
                      ^~~~~~~~~~~~
main.rs:14:12: 14:24 help: pass `--explain E0308` to see a detailed explanation
error: aborting due to 2 previous errors
```

But if I use the version on master branch, it works fine instead.

**Comment by jnicholls** [comment]
This won't be the case w/ mio 0.4+, where the std::net types are no longer being used and instead mio owns the types.

**Comment by carllerche** [comment]
Well, on the mio version on crates.io (0.3.x), `mio::tcp:: TcpListener` is an alias for `std::net::TcpListener` which is probably what is causing this error. The master branch changed this such that mio implements its own tcp types (and will be released as 0.4) which is probably why you are seeing these errors.

I do think that the rust error message could be improved, but I don't think there is an mio bug here per se. I'm going to close this for now, but if you disagree, feel free to re-open.
